nixops/resources/vault_approle.py

```diff
@@ -0,0 +1,218 @@
+    _reserved_keys = [ "vaultAddress", "roleName", "roleId", "bindSecretId"
+                        "secretIdBoundCidrs", "tokenBoundCidrs", "policies"
+                        "secretIdNumUses", "secretIdTtl", "tokenNumUses"
+                        "tokenMaxTtl", "tokenTtl", "period", "enableLocalSecretIds"
+                        "tokenType", "secretId", "cidrList"
+                      ]
+
+    @classmethod
+    def get_type(cls):
+        return "vault-approle"
+
+    def __init__(self, depl, name, id):
+        nixops.resources.DiffEngineResourceState.__init__(self, depl, name, id)
+        self.role_name = self._state.get('roleName', None)
+        self.role_id = self._state.get('roleId', None)
+        self.vault_address = self._state.get('vaultAddress', None)
+        self.secret_id = self._state.get('secretId', None)
+
+    def show_type(self):
+        s = super(VaultApproleState, self).show_type()
+        return s
+
+    @property
+    def resource_id(self):
+        path = '/v1/auth/approle/role/'
+        return self.vault_address + path + self.role_name if self.role_name else None
+
+    def get_definition_prefix(self):
+        return "resources.vaultApprole."
+
+    def get_physical_spec(self):
+        path = '/v1/auth/approle/role/'
+        return {'url': self.vault_address + path + self.role_name } if self.role_name else {}
+
+    def prefix_definition(self, attr):
+        return {('resources', 'vaultApprole'): attr}
+
+    def _connect(self, vault_token):
+        if self.vault_token:
+            pass
+        else:
+            self.vault_token = nixops.vault_common.initializeVault(vault_token=vault_token)
+        try:
+            header = {"X-Vault-Token": vault_token}
+            path = '/v1/auth/token/lookup-self'
+            remote_endpoint = self.vault_address + path
+            r = requests.get(remote_endpoint, headers=header, timeout=5)
+        except requests.exceptions.RequestException as e:
+            raise e.args[0]
+        if r.status_code != 200:
+            self.log("Vault token is invalid. Request returned {} status".format(status))
+            raise Exception(str(r.status_code) + ' ' + r.reason, r.json())
+        # log warning if token lifetime is less then 30 min ?!s
+        # check for permissions and take sure it can create role_ids
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+
+        # Deploying again won't change the role id. ..... 
+        config = self.get_defn()
+        self.vault_address = config['vaultAddress']
+        self.vault_token = config['vaultToken']
+        self.role_name = config['roleName']
+        self._connect(self.vault_token)
+
+        self.log("Creating Approle {}' at {}".format(self.role_name, self.vault_address))
+        try:
+            header = {"X-Vault-Token": self.vault_token}
+            path = '/v1/auth/approle/role/'
+            remote_endpoint = self.vault_address + path + self.role_name
+            data = {
```

Comment:
I would expect the API to support modifying most of the configs which I don't see being handled.
Which brings us back to the diff engine handers :) It should be much cleaner to define a handler for each modifiable attribute so that any change from the nix expression is handled correctly

Reply:
As i said in the description i will be taking care of that next.